keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,775 @@
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+#### No support JobSuccessPolicy for the NonIndexed Job
+
+As I described in [Non-Goals](#non-goals), we don't support the SuccessPolicy for the job with `NonIndexed` mode.
+
+#### Different between "complete" and "successCriteriaMet"
+The similar job conditions, `Complete` and `SuccessCriteriaMet`, are different in the following:
+
+- `Complete` means that all pods completed, not failed.
+- `SuccessCriteriaMet` means that the job meets the one of successPolicies.
+
+So, the job could have both conditions, `Complete` and `SuccessCriteriaMet`.
+
+#### The CronJob concurrentPolicy is not affected by JobSuccessPolicy
+Even after introducing the JobSuccessPolicy, all CronJob concurrentPolicies work as before
+since the JobSuccessPolicy doesn't change the semantics of the existing Job `Complete` condition 
+and the Job declares succeeded by a new condition, `SuccessCriteriaMet`.
+
+Specifically, the CronJob with `Forbid` concurrentPolicy created Jobs based on Job's `Complete` condition as before.
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.criteria.succeededIndexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// The successPolicy is immutable.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// Each criterion must have at least one acceptance criterion of
+	// "succeededIndexes" or "succeededCount".
+	// These criteria are evaluated in order, and the job meets one of the criteria,
+	// other criteria are ignored.
+	// After the job meets one of the criteria, the remaining running pods are removed.
+	//
+	// +optional
+	Criteria []SuccessPolicyCriterion
+}
+
+// SuccessPolicyCriteria describes a Job can be succeeded based on succeeded indexes.
```

Comment:
indicate that at least one of `SucceededIndexes` or `SucceededCount must be specified.

Reply:
Done.